walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -134,56 +156,173 @@
-      TimeRange nextEndingTimeRange = endOrderedTimeRanges[endOrderedPointer + 1];
-      while (nextEndingTimeRange.end() <= startingTimeRange.start()) {
-        endingTimeRange = nextEndingTimeRange;
-        endOrderedPointer++;
-        if (endOrderedPointer + 1 < endOrderedTimeRanges.length) {
-          nextEndingTimeRange = endOrderedTimeRanges[endOrderedPointer + 1];
-        } else {
+      // {@code TimeRange}, we can be certain that {@code endOrderedIndex + 1} won't go out
+      // of range.
+      while (endOrderedTimeRanges.get(endOrderedIndex + 1).end() <=
+          startOrderedTimeRanges.get(startOrderedIndex).start()) {
+        endOrderedIndex++;
+        if (!(endOrderedIndex + 1 < endOrderedTimeRanges.size())) {
           break;
         }
       }
 
       // At this point, we have:
       //    |----ending {@code TimeRange}----| available time |----starting {@code TimeRange}----|
-      int availableTimeRangeStart = endingTimeRange.end();
-      int availableTimeRangeEnd = startingTimeRange.start();
+      int availableTimeRangeStart = endOrderedTimeRanges.get(endOrderedIndex).end();
+      int availableTimeRangeEnd = startOrderedTimeRanges.get(startOrderedIndex).start();
       checkDurationAndAddAvailableTimeRange(requestedDuration,
                                             availableTimeRangeStart, availableTimeRangeEnd,
                                             availableTimeRanges);
 
-      if (startOrderedPointer + 1 < startOrderedTimeRanges.length &&
-          endOrderedPointer + 1 < endOrderedTimeRanges.length) {
-        startingTimeRange = startOrderedTimeRanges[++startOrderedPointer];
-        endingTimeRange = endOrderedTimeRanges[++endOrderedPointer];
+      if (startOrderedIndex + 1 < startOrderedTimeRanges.size() &&
+          endOrderedIndex + 1 < endOrderedTimeRanges.size()) {
+        startOrderedIndex++;
+        endOrderedIndex++;
       } else {
         return;
       }
     }
   }
 
+  /**
+   * Finds the overlapping, available {@code TimeRange} for mandatory and optional attendees. 
+   * If no {@code TimeRange} satisfies all mandatory and optional attendees, returns those
+   * {@code TimeRange} that satisfies mandatory attendees.
+   *
+   * @param requestedDuration The requested length of the overlapping, available {@code TimeRange}
+   *     to find.
+   * @param availableTimeRangesForMandatoryAttendees A list of available {@code TimeRange} for the
+   *     mandatory attendees. This list is expected to be sorted in ascending chronological order.
+   * @param availableTimeRangesForOptionalAttendees A list of available {@code TimeRange} for the
+   *     optional attendees. This list is expected to be sorted in ascending chronological order.
+   * @return A list of all available {@code TimeRange} for both mandatory and optional attendees,
+   *     sorted in ascending chronological order.
+   */
+  private List<TimeRange> getOverlappingAvailabeTimeRanges(long requestedDuration,
```

Comment:
Instead of doing all of this, how much would it hurt the time complexity to just call getConcernedAttendeesTimeRangesFromEvents(mandatory + optional), and if that's 0 then getConcernedAttendeesTimeRangesFromEvents(mandatory), and if that's 0 then getConcernedAttendeesTimeRangesFromEvents(optional)?

Reply:
The time complexity would remain the same. Sorting time ranges, which takes O(nlogn), is still the bottleneck.
Do you think this would make the code more readable?

Reply:
probably, since it'd p much be reusing code you wrote for the first pull request

Reply:
Got it, I'll make the change. Thank you so much for this!